### PR TITLE
Add a file describing how to contribute.

### LINKED ISSUE #2678: Have a CONTRIBUTE file
From this page: http://lwn.net/Articles/688070/

> projects ought to have a CONTRIBUTING file in their repository that describes the procedures to follow when a bug is found, when the user wants to fix a bug or add a feature, and when the user wants to ask a question. If a file named CONTRIBUTING exists in the repository, GitHub will provide a link to it along with a short message at the top of each new-issue page.

### PR DESCRIPTION
**Pull request by drwells**
I took a stab at it. Comments?

Closes #2678.

### CODE REVIEW

**Comment by drwells** [comment]
You can view the GitHub-highlighted version at

https://github.com/drwells/dealii/blob/add-contributing-file/CONTRIBUTING.md

**Comment by davydden** [comment]
we were discussing with @jppelteret the part related to 

> Asking and answering questions about the library

I think this section can be greatly expanded and be used as a "read this first" on Google group, where often people don't know how to ask questions and what type.

@jppelteret : perhaps you can give you draft to David and we polish it all together in this PR?

**Comment by drwells** [comment]
Sounds good. I will edit in the draft when you send it.

**Comment by jppelteret** [comment]
@drwells It appears that @davydden and I have opposing views on what he and I had discussed. 

My plan is to write (with the approval of all interested parties) a sticky post for the forum that expresses what the purpose of the forum is, how a user should pose questions there and which type of questions they can expect a meaningful answer to. It aims to point the users towards the documentation (and video lectures), describe the typical steps that new users might take to introduce themselves to deal.II (already mentioned in the documentation, I know...), and outline the expectation that some effort should into finding a solution to one's own problem before posting questions. 

Assuming that they've done an expected minimum amount of legwork and have not been able to solve their issue, one would next inform them what type of information we need in order to assist them in finding the root cause of their problem. This would serve to minimise the number of times one has to ask for a list of configuration options or the version of deal.II that they're running etc.. Ultimately, **my** real goal is to reduce the number of questions that, bluntly put, follow along the lines of

> Hi, I’m new. Please solve my not very well described problem, which I have not spent much time investigating, for me.

, and politely [disuade the **lazy**](http://lmgtfy.com/) from posting without applying their own mind and time to the problem.

I had actually started to write (but hadn't completed) an email to @bangerth to ask how he's responded to users who have contacted him directly (this surely must have happened in the past). I can expect that he has on many occasions handled them with the professionalism, patience and delicacy that their individual situations would have required. I'm curious as to whether he has a "template" response(s) to these questions that I can use as a basis for such a post.

Overall, I don't think that its appropriate to mention all of that here since it pertains in particular to the forum, and not deal.II as a whole. Once we've got such a post, perhaps we can post a link to it, and perhaps [the sticky](https://groups.google.com/forum/#!topic/dealii/uFi56KMx1h0) that provides other guidelines to using the forum?

As a final point, I'm once again left feeling ambivalent and self-conscious about my public reactions. Its important that its understood that I want to foster an environment for discussions about deal.II that is healthy for the new users of this great project, **as well as** more advanced users and the maintainers who spend an enormous amount of time helping everyone out. If such a post on the forum could be seen as too aggressive or might negatively change the way that people interact on the forum, then I'm alright if we leave things the way that they are.

**Comment by drwells** [comment]
@jppelteret I do not post very frequently in the forum so I will accept your judgement here (in case it has not been said recently, I really do appreciate the work you do on the mailing list :+1:). Perhaps we should discuss the proposed forum guide on the developer mailing list.

Fortunately, I think that for the sake of the `CONTRIBUTING.md` document we can keep things pretty vague and link to the forum sticky, so hopefully this is an easier problem than determining a mailing list policy.

**Comment by tjhei** [comment]
I really like the current PR and I would prefer if the CONTRIBUTE file is as succinct as possible without duplicating information but just points at it.

Regarding the mailing list, it might make sense to point at and/or extend https://www.dealii.org/mail.html too.

**Comment by tjhei** [comment]
There is also https://www.dealii.org/participate.html by the way.

**Comment by drwells** [comment]
Yes. I borrowed heavily from that.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+## Making changes to the library
+To make a change to deal.II you should create a *fork* (through GitHub) of the library and a separate *branch* (sometimes called feature branches). You can propose that your branch be combined with the rest of the library by opening a *pull request*. This will give a chance for others to review your code. While this seems very formal, keeping all of the code review in one place makes it easier to coordinate changes to the library (and there are usually several people making changes to the library at once). This process is described at length in [lecture 32.8](http://www.math.tamu.edu/~bangerth/videos.676.32.8.html). Please do not hesitate to ask questions about the workflow on the mailing list if you are not sure what to do.
+
+deal.II uses `astyle` with a specified configuration to ensure that the entire library is formatted in a consistent way. The use of `astyle` is documented [here](https://dealii.org/developer/index.html). The GNU style guide, which most editors support, is a decent approximation of the coding style used in deal.II.
```
Comment (tjhei) [inline CONTRIBUTING.md]: can you link to http://dealii.org/developer/doxygen/deal.II/CodingConventions.html instead?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Contributing to deal.II
+deal.II is a community project that lives by the participation of its members — i.e., including you! It is our goal to build an inclusive and participatory community so we are happy that you are interested in participating.
+
+## Getting started with git and GitHub
+If you are new to using git or the GitHub platform you may find it helpful to review [lecture 32.8](http://www.math.tamu.edu/~bangerth/videos.676.32.8.html), which should be enough to help you get started with using GitHub or possibly contributing to the library itself.
+
+## Asking and answering questions about the library
+The deal.II community maintains an active mailing list hosted by google groups [here](https://groups.google.com/forum/#!forum/dealii). The mailing list is for questions about the library of all levels.
```
Comment (tjhei) [inline CONTRIBUTING.md]: for more info see http://www.dealii.org/mail.html

File: CONTRIBUTING.md (hunk 1 above)
Comment (drwells) [inline CONTRIBUTING.md]: I thought about doing that but I am concerned that we might not want to hit new contributors with a whole page of style guidelines. I'll think about some way to phrase it to not sound exclusive.

**Comment by drwells** [comment]
@tjhei I made your recommended changes and added a link to the website's contributing page. I agree that we should try to keep this file succint so I like its current length.

@jppelteret Though I think this file is ready to go I would still like to hear what you have to say about participation on the mailing lists. Do you have your comments ready? Would you like to see the bit on the mailing list changed?

**Comment by tjhei** [comment]
Looks good to me as is!

> Though I think this file is ready to go I would still like to hear what you have to say about participation on the mailing lists.

Can we put that information into a sticky post on the google group instead? That way you will see it no matter how you find your way there.

**Comment by drwells** [comment]
That's a good point. I don't think we really need any more information on mailing list etiquette in this file so we can skip that (we can always add it in later).

**Comment by davydden** [comment]
i merged, if needed, let's update in another PR.

**Comment by jppelteret** [comment]
@drwells Thanks for your note of appreciation :-)

To you and @davydden I'm busy putting together a basic outline of an introductory post to the user google group. You can find it [here](https://gist.github.com/jppelteret/ca5a1396f4e4f8f78bf39a12820ed9eb) for now, and once I've finished gathering my thoughts I'll post it to the developer group for further discussion.

**Comment by Rombur** [comment]
@jppelteret what I have seen on some mailing lists, is a template of a question with fields like:
 - have you check the faq?
 - which version are using?
 - which compiler?
 - etc.

This avoids messages like "My code doesn't work. Please help."

**Comment by jppelteret** [comment]
Thanks @Rombur! I've updated [my notes](https://gist.github.com/jppelteret/ca5a1396f4e4f8f78bf39a12820ed9eb) to reflect this. 

I think that my list of points to consider is nearly complete. I'll start to flesh this out more thoroughly and will hopefully have a proposal by the end of the weekend.

**Pull request merged**